group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorService.java

```diff
@@ -0,0 +1,581 @@
+
+    public static class Builder {
+        private final int nodeId;
+        private final GroupCoordinatorConfig config;
+        private PartitionWriter<Record> writer;
+        private CoordinatorLoader<Record> loader;
+
+        public Builder(
+            int nodeId,
+            GroupCoordinatorConfig config
+        ) {
+            this.nodeId = nodeId;
+            this.config = config;
+        }
+
+        public Builder withWriter(PartitionWriter<Record> writer) {
+            this.writer = writer;
+            return this;
+        }
+
+        public Builder withLoader(CoordinatorLoader<Record> loader) {
+            this.loader = loader;
+            return this;
+        }
+
+        public GroupCoordinatorService build() {
+            if (config == null)
+                throw new IllegalArgumentException("Config must be set.");
+            if (writer == null)
+                throw new IllegalArgumentException("Writer must be set.");
+            if (loader == null)
+                throw new IllegalArgumentException("Loader must be set.");
+
+            String logPrefix = String.format("GroupCoordinator id=%d ", nodeId);
+            LogContext logContext = new LogContext(String.format("[%s ]", logPrefix));
+
+            CoordinatorBuilderSupplier<ReplicatedGroupCoordinator, Record> supplier = () ->
+                new ReplicatedGroupCoordinator.Builder(config);
+
+            CoordinatorEventProcessor processor = new MultiThreadedEventProcessor(
+                logContext,
+                "group-coordinator-event-processor-",
+                config.numThreads
+            );
+
+            CoordinatorRuntime<ReplicatedGroupCoordinator, Record> runtime =
+                new CoordinatorRuntime.Builder<ReplicatedGroupCoordinator, Record>()
+                    .withLogPrefix(logPrefix)
+                    .withLogContext(logContext)
+                    .withEventProcessor(processor)
+                    .withPartitionWriter(writer)
+                    .withLoader(loader)
+                    .withCoordinatorBuilderSupplier(supplier)
+                    .build();
+
+            return new GroupCoordinatorService(
+                logContext,
+                config,
+                runtime
+            );
+        }
+    }
+
+    /**
+     * The logger.
+     */
+    private final Logger log;
+
+    /**
+     * The group coordinator configurations.
+     */
+    private final GroupCoordinatorConfig config;
+
+    /**
+     * The coordinator runtime.
+     */
+    private final CoordinatorRuntime<ReplicatedGroupCoordinator, Record> runtime;
+
+    /**
+     * Boolean indicating whether the coordinator is active or not.
+     */
+    private final AtomicBoolean isActive = new AtomicBoolean(false);
+
+    /**
+     * The number of partitions of the __consumer_offsets topics. This is provided
+     * when the component is started.
+     */
+    private volatile int numPartitions = -1;
```

Comment:
just to confirm this is usually just OffsetsTopicPartitionsProp, but we set to 1 in tests etc?

Reply:
The __consumer_offsets is created based on `OffsetsTopicPartitionsProp`. However, `numPartitions` is gotten from the metadata cache when the coordinator is started. If the topic was manually created, it could have a different number of partitions.

Reply:
`numPartitions` is set in `startup` all the time.

Reply:
Sorry, I mean the topic should only be manually created via tests. Usually it would be done automatically when we use consumers.